### ISSUE TITLE
A long delay in auto frame rate matching

### ISSUE THREAD

**Issue by Bec-de-Xorbin** [moneytoo/Player#632]
**Describe the bug**
There's a long delay, usually about 10 to 20 seconds, before display matches with a video frame rate. This started in v0.182 when an FFmpeg dependency was removed. Before it was pretty much instant.

**Device (please complete the following information):**
 - Model: Chromecast with Google TV HD
 - OS: Android 12
 - Version 0.183

**Comment by moneytoo** [comment]
The app needs to read enough video track samples to get the average frame rate. I set it to 1000 samples to get good enough precision. But I can lower it to 350 which is just above the limit if I look for commonly used frame rates (59.94 vs 60 fps). Let's try version from https://brouken.com/player/ if it still works as expected. Hopefully it should get rid of almost 2/3 of the delay.

**Comment by Bec-de-Xorbin** [comment]
It's now quite good on most files taking about 2-5 seconds but still not as good as 0.181 obviously. One 23.976 fps movie I tried to play still takes 10 to 15 seconds until it switches the frame rate. I also thought you removed auto-play and that's why I manually pressed play right away but it seems that when it gets the frame rate it will play back automatically so that makes the issue a bit less severe.

Was the FFmpeg dependency too laborious to maintain or why was it removed?

**Comment by moneytoo** [comment]
> Was the FFmpeg dependency too laborious to maintain or why was it removed?

FFmpeg is available as ready made library so maintenance wasn't a problem. The problem was size and principle. Without FFmpeg, the app is only 15 MB while adding FFmpeg, useful only for frame rate detection additional 55 MB. It packs all the decoders/encoders and it wasn't only ffprobe. It could probably be trimmed down, but that would be additional burden.

All of this shouldn't be needed as modern Android (12+) devices can change refresh rates without switching display modes. ExoPlayer already uses by default the `CHANGE_FRAME_RATE_ONLY_IF_SEAMLESS` strategy and it indeed changes display refresh rate on my devices. I could try using `CHANGE_FRAME_RATE_ALWAYS` strategy instead. Would you be willing to test that if I prepare a version with that?

**Comment by moneytoo** [comment]
Try https://brouken.com/player/frame_rate_strategy/ Does it switches display refresh rate at all? It ignores auto frame rate matching settings and attempts to set `CHANGE_FRAME_RATE_ALWAYS` strategy.

**Comment by moneytoo** [comment]
I managed to upgrade my Xiaomi Mi Box S to Android 12. Seems like that the `CHANGE_FRAME_RATE_ALWAYS` strategy works and fully switches display modes. I will add a switch for all the possible options.

**Comment by Bec-de-Xorbin** [comment]
> Try https://brouken.com/player/frame_rate_strategy/ Does it switches display refresh rate at all? It ignores auto frame rate matching settings and attempts to set `CHANGE_FRAME_RATE_ALWAYS` strategy.

No, it doesn't switch display refresh rate at all.

**Comment by moneytoo** [comment]
Yep. In my case the display was switching because of HDR and not refresh rate. Seems like the `CHANGE_FRAME_RATE_ALWAYS` strategy is not supported in ExoPlayer so currently used switching is still preferrable (as it's less work and has the option to switch resolution as well).
